feat: backpack

pabloes opened the pull request:
Starts https://github.com/dclexplorer/bevy-ui-scene/issues/9

# Stage 1 - Wearables:

![Image](https://github.com/user-attachments/assets/c14faa08-61f3-4d62-ac4f-1ddf15537304)

- [ ] Avatar Preview
  - [x] R2T
  - [ ] Avatar rotate
  - [ ] Zoom in/out
  - [x] Pan

- [ ] Wearable Category
  - [x] Hover/Clicked highlights
  - [ ] Sound effects
  - [x] Focus category area on the Avatar Preview (Head, Body, Shoes) **(DONE but pendent to review current unity explorer version new camera positions)**

-  [ ] Wearable List
    -  [x] Hover/Clicked highlights
    -  [ ] Sound effects
    -  [x] Category filters
    -  [x] Rarity icon, border and color
    -  [x] Equip/unequip actions

-  [ ] Wearable Info
    - [x] Name, type, image, description
    - [ ] Hide info **(ADDED)**

- [ ] Color Selector
    - [ ] For body, head, hair, beard, eyes and eyebrows colour

(reference for color selector)
![Image](https://github.com/user-attachments/assets/2aa98e02-ac32-4292-97bd-52a5b475b2bc)

pabloes committed:
feat: backpack in progress

pabloes committed:
clean console.log

pabloes committed:
feat: backpack: catalog navbar

pabloes committed:
fix: backpack: catalog navbar rightArrow relative size

pabloes committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into feat/backpack-merge

# Conflicts:
#	scene/assets/images/atlas/backpack.json
#	scene/assets/images/atlas/backpack.png
#	scene/package-lock.json
#	scene/package.json
#	scene/src/components/button-text-icon/ButtonTextIcon.tsx
#	scene/src/json/backpack-data.ts
#	scene/src/state/store.ts
#	scene/src/state/types.ts
#	scene/src/ui-classes/main-menu/MainMenu.tsx

pabloes committed:
fix: fixes after merge from main

pabloes committed:
fix: menu button sizes

pabloes committed:
feat: backpack: wearable catalog sections grid + selected item

pabloes committed:
feat: add URN type definition

pabloes committed:
feat: add URN type definition

pabloes committed:
feat: loading wearable sprite
+ nft-empty sprite

pabloes committed:
feat: backpack: init and fetch wearables page

pabloes committed:
feat: backpack: catalog: use rarity backgrounds once page data is loaded

pabloes committed:
feat: backpack: use player/guest address/userId for fetching wearables

pabloes committed:
fix type

pabloes committed:
feat: backpack: wip loading player outfit

pabloes committed:
feat: backpack: set rarity backgroun on outfit wearable list component

pabloes committed:
refactor: wearables-promise-utils.ts

pabloes committed:
refactor: type definition WearablesPageResponse

pabloes committed:
feat: cache catalystWearableMap also from catalog pages

pabloes committed:
refactor: backpackPage react elements

pabloes committed:
refactor: backpackPage react elements

pabloes committed:
revert last 2 commits refactor:
causes flickering issue and layout miss-alignment

pabloes committed:
feat: backpack: wip: pagination component

pabloes committed:
feat: backpack: pagination

pabloes committed:
feat: backpack: pagination: disabled when loadingPage

pabloes committed:
feat: backpack: filter by category

pabloes committed:
feat: backpack: category change resets currentPage to 1

pabloes committed:
fix: backpack: fixBodyCategoryURIParam

pabloes committed:
refactor: backpack: react containers

pabloes committed:
fix: backpack: fix outfit component

pabloes committed:
fix: backpack: fix outfit component

pabloes committed:
feat: backpack: wearables-catalog: show orange border on equipped items

pabloes committed:
feat: backpack: wearables-catalog: add selection frame to spritesheet

pabloes committed:
feat: backpack: wearables-catalog: add selection frame to spritesheet

pabloes committed:
feat: backpack: wearables-catalog: selectable wearable cells

pabloes committed:
feat: backpack: wearables-catalog: selection: add button

pabloes committed:
feat: backpack: wearables-catalog: selection: info-panel WIP

pabloes committed:
feat: backpack: wearables-catalog: selection: info-panel WIP

pabloes committed:
feat: backpack: wearables-catalog: selection: info-panel WIP: allow unselect

pabloes committed:
feat: backpack: wearables-catalog: selection: info-panel WIP: details

pabloes committed:
feat: backpack: wearables-catalog: WIP equip/unequip

pabloes committed:
feat: backpack: wearables-catalog: WIP

pabloes committed:
feat: backpack: wearables-catalog WIP: hide selection frame when loading page

pabloes committed:
feat: backpack: wearables-catalog WIP: equip/unequip

pabloes committed:
feat: backpack: wearables-catalog WIP: fix base wearables name

pabloes committed:
feat: backpack: wearables-catalog WIP: InfoPanel tweaks

pabloes committed:
feat: backpack: wearables-catalog WIP: calling setAvatar

pabloes committed:
feat: backpack: wearables-catalog WIP: use URN with tokenId on equippedWearables and outfitSetup

pabloes committed:
fix: mainhud icons

pabloes committed:
fix: backpack: wearables-catalog WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: avatar-preview WIP

pabloes committed:
feat: backpack: fix base_body related issues

pabloes committed:
feat: backpack: avatar preview as absolute bellow catalog

pabloes committed:
feat: backpack: avatar preview: set camera position per category
+ fix body_shape category

pabloes committed:
feat: backpack: avatar preview: set camera position per category
+ fix body_shape category

pabloes committed:
feat: backpack: // TODO when  loading, don't let change any filter : category

pabloes committed:
feat: backpack: // TODO cleaning

pabloes committed:
refactor: // TODO use constant ROUNDED_TEXTURE_BACKGROUND

pabloes committed:
refactor: clean // TODO

pabloes committed:
refactor: rename type CatalystWearable to WearableEntity

pabloes committed:
fix: backpack: handling empty equippedWearables

pabloes committed:
refactor: backpack: move getURNWithoutTokenId to utils/URN-utils.ts

pabloes committed:
refactor: backpack: move getURNWithoutTokenId to utils/URN-utils.ts

pabloes committed:
test: backpack: add unit test for getURNWithoutTokenId

pabloes committed:
test: backpack: fix unit test for getURNWithoutTokenId

pabloes committed:
test: backpack: fix unit test for getURNWithoutTokenId

pabloes committed:
test: backpack: WIP outfit service tests

pabloes committed:
fix: backpack: WIP fix mutation on EMPTY_OUTFIT

pabloes committed:
fix: backpack: WIP fix mutation on EMPTY_OUTFIT

pabloes committed:
prettier format

pabloes committed:
feat: backpack: InfoPanel: // TODO create component <Tag />

pabloes committed:
feat: backpack: catalog grid // TODO improve image rounded-border.png

pabloes committed:
feat: backpack WIP: catalog grid fix equipped border

pabloes committed:
feat: backpack WIP: clean TODO

pabloes committed:
feat: backpack WIP: URNWithoutTokenId // TODO improve to specific strings

pabloes committed:
feat: backpack WIP:  // TODO use realm BaseURL

pabloes committed:
feat: backpack WIP: // TODO use realm BaseURL

pabloes committed:
feat: backpack WIP: fix grid equipped

pabloes committed:
feat: backpack WIP: fix setAvatar expects urns with tokenId

pabloes committed:
feat: backpack WIP: lint

pabloes committed:
feat: backpack WIP: // TODO clean + type WearableRatity

pabloes committed:
feat: backpack WIP: refactor

pabloes committed:
feat: backpack WIP: clean console.log

pabloes committed:
feat: backpack WIP: refactor

pabloes committed:
feat: backpack: avatar preview: send wearables urn with tokenId

pabloes committed:
feat: backpack: WIP: add TODO

pabloes committed:
feat: backpack: WIP: add viewport store

pabloes committed:
feat: backpack: WIP: canvas getContentWidth & getContentHeight

pabloes committed:
feat: backpack: WIP: fix miss-alignments

pabloes committed:
feat: backpack: WIP: fix miss-alignments

pabloes committed:
feat: backpack: WIP: fix miss-alignments

pabloes committed:
feat: backpack: WIP: implement redux in BackpackPage state

pabloes committed:
feat: backpack: WIP: clean TODO remove auto-open backpack

pabloes committed:
feat: backpack: WIP: add TODO

pabloes committed:
feat: backpack: WIP: remove   // eslint-disable-next-line @typescript-eslint/ban-types

pabloes committed:
feat: backpack: WIP: fix saveAvatar

pabloes committed:
feat: backpack: WIP: fix format and lint

kuruk-mm committed:
format

kuruk-mm commented on scene/src/bevy-api/interface.ts:
```diff
@@ -25,7 +27,25 @@ export type KernelFetchRespose = {
+  equip?: PBAvatarEquippedData
+}
 export type BevyApiInterface = {
+  setAvatar: (avatarData: SetAvatarData) => Promise<any> // TODO set profile type
```
```suggestion
  setAvatar: (avatarData: SetAvatarData) => Promise<number> // TODO set profile type
```

if I´m not wrong. The Bevy implementation returns the deployed version, which it is a number
cc @robtfm

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/d0b74fa20245f12fcc2f833e257d1ef5ea3301f7

kuruk-mm commented on scene/src/components/backpack/AvatarPreview.tsx:
```diff
@@ -0,0 +1,135 @@
+  AvatarShape.getMutable(avatarPreview.avatarEntity).eyeColor =
+    avatarBase.eyesColor
+  AvatarShape.getMutable(avatarPreview.avatarEntity).skinColor =
+    avatarBase.skinColor
```
can you put these into a variable?

```
const avatarShape = AvatarShape.getMutable(avatarPreview.avatarEntity)
avatarShape.wearables = ...
```

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/bf9cb3587740a15b5aae804de24d53125eb75c8b

pabloes committed:
feat: backpack: WIP: fix format and lint

kuruk-mm commented on scene/src/components/backpack/InfoPanel.tsx:
```diff
@@ -0,0 +1,172 @@
+        >
+          <UiEntity
+            uiTransform={{
+              width: canvasScaleRatio * 400,
```
would be nice to avoid nesting, try to use

```typescript
if (!wearable) return null

return (
  <UiEntity>
    <UiEntity>...</UiEntity>
  </UiEntity>
)
```

or also, you can use:

```
{wearable && (
  <UiEntity>
    <UiEntity>...</UiEntity>
  </UiEntity>
)}
```

instead of inline-if

pabloes replied:
changed ESLINT rules
https://github.com/dclexplorer/bevy-ui-scene/commit/f602fde4e8abc77eb675de49eb50b9a169d82e10

kuruk-mm commented on scene/src/components/backpack/WearableCatalogGrid.tsx:
```diff
@@ -0,0 +1,317 @@
+  memo: Record<URNWithoutTokenId, boolean> // TODO consider using Map if possible for performance improvement because long keys
+} = {
+  equippedWearables: [],
+  memo: {},
```
```suggestion
  memo: new Map<URNWithoutTokenId, boolean>(),,
```

would be great to use a Map here, and use the functions `has(), get(), includes(), set()`, and also it is well-defined

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/f5dcdfe02e7ac3ab1f17985e7e20fcab4cd14407

kuruk-mm commented on scene/src/components/backpack/WearableCatalogGrid.tsx:
```diff
@@ -0,0 +1,317 @@
+              )
+              onChangeSelection(state.selectedWearableURN)
+            }}
+          >
```
This is OK, but creating a `WearableCatalogItem` can be interesting to have it better sub-divided.

And you can also defines differents behaviors while is loading or not, while it's there, avoiding so many `!loading &&`

Up to you to modify it 🙌

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/6b809c03e9292c7f9a514ee494f9564a070b0eba
Refactor is there but I note !loading issue for next branch

kuruk-mm commented on scene/src/components/backpack/WearableCatalogGrid.tsx:
```diff
@@ -0,0 +1,317 @@
+                loading={loading}
+              />
+            ) : null}
+            {state.selectedWearableURN === wearableElement?.urn && !loading ? (
```
try to avoid if-nesting with `null`, this could be
```
{state.selectedWearableURN === wearableElement?.urn && !loading && (
  <UiEntity>
   .....
  </UiEntity>
)}
```

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/f07b05588e391e642e2757da44745eeefe0bd460

after previous ESLINT rules change

kuruk-mm commented on scene/src/components/backpack/WearableCategoryButton.tsx:
```diff
@@ -0,0 +1,125 @@
+      }}
+      onMouseDown={() => {
+        callbacks.onClick()
+      }}
```
I think this can be done, and looks simpler, without the need of the `callbacks` variable

```suggestion
  onMouseDown={() => (onClick ?? noop)()}
```

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/5ccb06305414633916aff1b02c6259eba457da45

kuruk-mm commented on scene/src/components/backpack/WearableCategoryButton.tsx:
```diff
@@ -0,0 +1,125 @@
+          positionType: 'absolute',
+          position: {
+            left: 56 * canvasScaleRatio * 2
+          }
```
I see that there is a lot of `canvasScaleRatio * 2`, could you use a const like `const scaleRatio = canvasScaleRatio * 2`?

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/89bfd8fe37c57e2105b7c414f5e82b025c7791cd

kuruk-mm commented on scene/src/components/backpack/WearableCategoryButton.tsx:
```diff
@@ -0,0 +1,125 @@
+            src: `https://peer.decentraland.org/lambdas/collections/contents/${selectedURN}/thumbnail`
+          },
+          textureMode: 'stretch'
+        }
```
small change: i think inverting to avoid using `selectedURN === null`, is more redeable
but if you prefer this way, I don't see issues about it
```suggestion
const textureProps: UiBackgroundProps = selectedURN
  ? {
      texture: {
        src: `https://peer.decentraland.org/lambdas/collections/contents/${selectedURN}/thumbnail`
      },
      textureMode: 'stretch'
    }
  : getBackgroundFromAtlas({
      atlasName: 'backpack',
      spriteName: 'nft-empty'
    })
```

pabloes replied:
ESLint: Unexpected nullable string value in conditional. Please handle the nullish/ empty cases explicitly.(@typescript-eslint/ strict-boolean-expressions)

https://github.com/dclexplorer/bevy-ui-scene/commit/cc99157f68ecea3fdcf00d3a447e6b75e15cc4e0

kuruk-mm commented on scene/src/components/backpack/WearableCategoryList.tsx:
```diff
@@ -0,0 +1,203 @@
+            onSelectCategory(activeCategory === 'skin' ? null : 'skin')
+          }}
+          selectedURN={outfitSetup.wearables.skin}
+        />
```
could be interesting doing a map here from a categories array like:

```typescript
const CATEGORIES: WearableCategory[] = [
  'body_shape', 'eyebrows', 'mouth', 'upper_body', 'lower_body', 'hat',
  'earring', 'tiara', 'helmet', 'hair', 'eyes', 'facial_hair',
  'hands_wear', 'feet', 'eyewear', 'mask', 'top_head', 'skin'
]

{CATEGORIES.map((category) => (
  <WearableCategoryButton
    key={category}
    category={category}
    active={activeCategory === category}
    onClick={() => onSelectCategory(activeCategory === category ? null : category)}
    selectedURN={outfitSetup.wearables[category]}
  />
))}
```

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/3b1b4ece043afcfc297bc175bf69386c2809522f

kuruk-mm commented on scene/src/components/nav-button/NavButton.tsx:
```diff
@@ -0,0 +1,96 @@
+      }}
+      onMouseDown={() => {
+        callbacks.onClick()
+      }}
```
```suggestion
      onMouseDown={onClick}
```

this can be simplified as that, if I'm not wrong

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/da857c8d4cf4e24deef37154d2b0e0936e1bedfa

kuruk-mm commented on scene/src/components/nav-button/NavButton.tsx:
```diff
@@ -0,0 +1,96 @@
+        <UiEntity
+          onMouseDown={() => {
+            callbacks.onDelete()
+          }}
```
```suggestion
          onMouseDown={onDelete}
```

same

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/da857c8d4cf4e24deef37154d2b0e0936e1bedfa

kuruk-mm commented on scene/src/components/nav-button/NavButton.tsx:
```diff
@@ -0,0 +1,96 @@
+        callbacks.onClick()
+      }}
+    >
+      {icon != null ? (
```
this can be changed to
{ icon && <Icon\> }

pabloes replied:
ESLint: Unexpected nullable object value in conditional. An explicit null check is required.(@typescript-eslint/ strict-boolean-expressions)

But `{Boolean(icon) && (` looks fine

pabloes replied:
after eslint rules change

https://github.com/dclexplorer/bevy-ui-scene/commit/0a36aaef5588c0bb152d8b6c2c9816abba1aceeb

kuruk-mm commented on scene/src/scenes/main.tsx:
```diff
@@ -10,6 +10,8 @@ let gameInstance: GameController
   gameInstance = new GameController()
   gameInstance.uiController.loadingAndLogin.startLoading()
+
+  // BevyApi.loginGuest()
```
```suggestion
```

kuruk-mm commented on scene/src/utils/URN-utils.ts:
```diff
@@ -0,0 +1,46 @@
+import type { URN, URNWithoutTokenId } from './definitions'
```
can you rename this class to urn-utils.ts?

kuruk-mm commented on scene/src/utils/URN-utils.ts:
```diff
@@ -0,0 +1,46 @@
+export const BASE_MALE_URN: URNWithoutTokenId =
+  'urn:decentraland:off-chain:base-avatars:BaseMale' as URNWithoutTokenId
+
+const urnWithoutTokenIdMemo: Record<URN, URNWithoutTokenId> = {} // TODO consider using Map if possible for performance improvement because long keys
```
I agree with the comment. Maps will be better.

kuruk-mm commented on scene/src/utils/URN-utils.ts:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  // Check if it's an off-chain URN or has less than 6 parts
+  if (urn.includes(':off-chain:') || urn.split(':').length < 6) {
```
on checks like this, is great to put functions with the name of the action that you are going to do... so the code explains by itself... like

```typescript
function isOffChainOrIncomplete(urn: string): boolean {
  return urn.includes(':off-chain:') || urn.split(':').length < 6
}
```

and the result looks more clear

```typescript
if (isOffchainOrIncomplete(urn)) { /* ... */ }

kuruk-mm commented on scene/src/utils/dcl-utils.ts:
```diff
@@ -0,0 +1,4 @@
+import { timers } from '@dcl-sdk/utils'
+
+export const dclSleep = async (delay: number): Promise<any> =>
```
```suggestion
export const delay = async (delay: number): Promise<any> =>
```

can be part of the utils.ts, no need for adding the dcl on the front of the name function

pabloes replied:
this name is because it's using `import { timers } from '@dcl-sdk/utils`
so we don't confuse it with a classic sleep with native setTimeout that could be used at some moment in other cases in the project, e.g. maybe tests

btw, this function is not really used, it's a leftover now, I can delete it and add it again if necessary

kuruk-mm commented on scene/src/utils/function-utils.ts:
```diff
@@ -0,0 +1,17 @@
+export function noop(): void {}
+
+export function memoize<T, U>(fn: (arg: T) => U): (arg: T) => U {
```
nice!

kuruk-mm commented on scene/src/utils/wearables-promise-utils.ts:
```diff
@@ -0,0 +1,118 @@
+  pageSize,
+  wearableCategory,
+  address
+}: any): Promise<WearablesPageResponse> {
```
```suggestion
}: WearableCatalogPageParams): Promise<WearablesPageResponse> {
```

can we try to avoid the `any` here?

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/cae232fb9cd7ddabbe91374a0f23f5d5e55469a1#diff-44430281a4df0b2fc3bfb512bd4c191a352664a9c4b4ae8c1abf0c11499fe73d

kuruk-mm commented on scene/src/utils/wearables-promise-utils.ts:
```diff
@@ -0,0 +1,118 @@
+    throw error
+  }
+
+  function getWearableCatalogPageURL({
+    pageNum,
+    pageSize,
+    address,
+    wearableCategory,
+    includeBase,
+    includeOnChain,
+    catalystBaseUrl
+  }: WearableCatalogPageParams): string {
+    let str: string = `${catalystBaseUrl}/explorer/${address}/wearables?pageNum=${pageNum}&pageSize=${pageSize}&includeEntities=true`
+    if (wearableCategory !== null) {
+      str += `&category=${wearableCategory}`
+    }
+    if (includeBase) {
+      str += `&collectionType=base-wearable`
+    }
+    if (includeOnChain) {
+      str += `&collectionType=on-chain`
+    }
+    return str
+  }
+}
```
```suggestion
function getWearableCatalogPageURL(params: WearableCatalogPageParams): string {
  const { pageNum, pageSize, address, wearableCategory, includeBase, includeOnChain, catalystBaseUrl } = params
  let url = `${catalystBaseUrl}/explorer/${address}/wearables?pageNum=${pageNum}&pageSize=${pageSize}&includeEntities=true`

  if (wearableCategory) url += `&category=${wearableCategory}`
  if (includeBase) url += `&collectionType=base-wearable`
  if (includeOnChain) url += `&collectionType=on-chain`

  return url
}
```

minor suggestion

pabloes replied:
to avoid some !== null maybe we can remove ESLINT rule:
ESLint: Unexpected value in conditional. A boolean expression is required.(@typescript-eslint/ strict-boolean-expressions)

pabloes replied:
https://github.com/dclexplorer/bevy-ui-scene/commit/fd4f25f04b5fd2f9b1ac783216926cbea635feba

https://github.com/dclexplorer/bevy-ui-scene/commit/4960b577ecab96c6ab8c2a232ca07b187bf3226f

kuruk-mm reviewed (changes requested):
Great work! This is amazing! Thanks for what you are doing 💪 

I left small changes to the code, and some functionality issues that I had:

- The left/right markers are not working

- Empty sections should show a message
![image](https://github.com/user-attachments/assets/ec60c43a-6bdb-4f84-a996-6331d1f53e4d)


- No selected item should show it that label
![image](https://github.com/user-attachments/assets/76463214-002e-4331-9a73-1a88ba0a1252)


- Bug on pagination, when I change to the last tab (video attached)
https://github.com/user-attachments/assets/c44781b7-709a-4a31-a251-86100f62b8b5

- After you change the category, it should CLEAN immediately and then load whatever it needs to load


- The category icon in the thumbnail is missing
![image](https://github.com/user-attachments/assets/cf0d4f19-de5d-44b2-9b0f-ee75f18a2211)

pabloes committed:
Update scene/src/scenes/main.tsx

Co-authored-by: Mateo Miccino <mateomiccino@gmail.com>

pabloes committed:
feat: backpack: WIP: fix pagination and add unit test: when currentPage is 4 and pages are 4 , should return all items from 1 to 4

pabloes committed:
feat: backpack WIP: PR review: refactor AvatarPreview.tsx mutableAvatarShape in a variable

pabloes committed:
Merge remote-tracking branch 'origin/feat/backpack-merge' into feat/backpack-merge

pabloes committed:
Merge remote-tracking branch 'origin/feat/backpack-merge' into feat/backpack-merge

pabloes commented:
fixed pagination bug and added unit test here: https://github.com/dclexplorer/bevy-ui-scene/commit/24bf4f06cf8e8c81eadd554c7b64b3fa30351c58#diff-342f78abfb09d409fc2c92ca791e4b8ed0e05e229e7daec908beb9614c4bbd86

pabloes committed:
Merge remote-tracking branch 'origin/feat/backpack-merge' into feat/backpack-merge

pabloes committed:
feat: backpack WIP: PR review: BevyApiInterface setAvatar returns type number

pabloes committed:
feat: backpack WIP: PR review: change eslint rules null condition

pabloes committed:
feat: backpack WIP: PR review: WearableCatalogGrid change Record to Map

pabloes committed:
feat: backpack WIP: PR review: WearableCatalogGrid refactor adding WearableCatalogItem

pabloes committed:
feat: backpack WIP: PR review: WearableCatalogGrid refactor adding WearableCatalogItem

pabloes committed:
Merge remote-tracking branch 'origin/feat/backpack-merge' into feat/backpack-merge

pabloes committed:
feat: backpack WIP: PR review: WearableCatalogGrid refactor adding WearableCatalogItem

pabloes committed:
feat: backpack WIP: PR review: refactor to avoid null check (after previous eslint rules change)

pabloes committed:
feat: backpack WIP: PR review: refactor WearableCategoryButton.tsx remove callbacks object

pabloes committed:
feat: backpack WIP: PR review: refactor WearableCategoryButton.tsx  62 * canvasScaleRatio * 2

pabloes committed:
feat: backpack WIP: PR review: refactor WearableCategoryButton.tsx allowNullableString: true

pabloes committed:
feat: backpack WIP: PR review: refactor WearableCategoryList.tsx

pabloes committed:
feat: backpack WIP: PR review: refactor NavButton.tsx

pabloes committed:
feat: backpack WIP: PR review: refactor NavButton.tsx

pabloes committed:
feat: backpack WIP: PR review: rename urn-utils.ts

kuruk-mm commented on scene/src/components/backpack/WearableCatalogGrid.tsx:
```diff
@@ -0,0 +1,317 @@
+                      if (
+                        isEquipped(wearableElement, equippedWearables, baseBody)
+                      ) {
+                        void onUnequipWearable(wearableElement)
```
```suggestion
                        onUnequipWearable(wearableElement).catch(console.error)
```

pabloes replied:
fixed outside and defined the callbacks results as sync void:
```
         onUnequipWearable={(wearable: CatalogWearableElement): void => {
                  this.updateEquippedWearable(wearable.category, null)
                    .catch(console.error)
                }}
```

pabloes committed:
feat: backpack WIP: PR review: refactor urn-utils + record/map

pabloes committed:
feat: backpack WIP: PR review: refactor urn-utils + record/map

pabloes committed:
feat: backpack WIP: PR review: fix format & lint

pabloes committed:
feat: backpack WIP: PR review: refactor wearable-promise-utils.ts

pabloes committed:
feat: backpack WIP: PR review: fix WearableCatalogGrid.tsx callback type results

pabloes committed:
feat: backpack WIP: PR review: fix WearableCatalogGrid.tsx callback type results

pabloes committed:
feat: backpack WIP: PR review: refactor getWearableCatalogPageURL

pabloes committed:
feat: backpack WIP: PR review: refactor getWearableCatalogPageURL

pabloes committed:
feat: backpack WIP: PR review: refactor getWearableCatalogPageURL

kuruk-mm commented on scene/src/components/pagination/pagination-util.ts:
```diff
@@ -0,0 +1,21 @@
+          total - PAGE_BUTTONS + 1
+        )
+      : 0
+  console.log('ofset', offset)
```
Please remove log

pabloes committed:
feat: backpack WIP: PR review: remove console.log
